ui/StatusQ/src/leftjoinmodel.cpp

```diff
@@ -0,0 +1,219 @@
+    auto rightRoleNames = m_rightModel->roleNames();
+
+    if (!leftRoleNames.count() ||  !rightRoleNames.count()) {
+        qWarning() << "Both left and right models have to contain some roles!";
+        return;
+    }
+
+    auto leftModelJoinRoleList = leftRoleNames.keys(m_joinRole.toUtf8());
+    auto rightModelJoinRoleList = rightRoleNames.keys(m_joinRole.toUtf8());
+
+    if (leftModelJoinRoleList.size() != 1
+            || rightModelJoinRoleList.size() != 1) {
+        qWarning().noquote() << QString("Both left and right models have to "
+                                        "contain join role %1!").arg(m_joinRole);
+        return;
+    }
+
+    m_leftModelJoinRole = leftModelJoinRoleList.at(0);
+    m_rightModelJoinRole = rightModelJoinRoleList.at(0);
+
+    auto leftRoles = leftRoleNames.keys();
+    auto maxLeftRole = std::max_element(leftRoles.cbegin(), leftRoles.cend());
+    auto rightRolesOffset = *maxLeftRole + 1;
+    auto roleNames = leftRoleNames;
+
+    auto i = rightRoleNames.constBegin();
+    while (i != rightRoleNames.constEnd()) {
+        if (i.value() != m_joinRole) {
+            auto roleWithOffset = i.key() + rightRolesOffset;
+            roleNames.insert(roleWithOffset, i.value());
+            m_joinedRoles.append(roleWithOffset);
+        }
+        ++i;
+    }
+
+    m_rightModelRolesOffset = rightRolesOffset;
+    m_roleNames = roleNames;
+
+    connect(m_rightModel, &QAbstractItemModel::dataChanged, this,
+            [this](auto& topLeft, auto& bottomRight, auto& roles) {
+        QVector<int> rolesTranslated;
+
+        if (roles.contains(m_rightModelJoinRole)) {
+            rolesTranslated = m_joinedRoles;
+        } else {
+            rolesTranslated = roles;
+
+            for (auto& role : rolesTranslated)
+                role += m_rightModelRolesOffset;
+        }
+
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), rolesTranslated);
+    });
+
+    disconnect(m_leftModel, &QAbstractItemModel::rowsInserted,
+               this, &LeftJoinModel::initializeIfReady);
+    disconnect(m_rightModel, &QAbstractItemModel::rowsInserted,
+               this, &LeftJoinModel::initializeIfReady);
+
+    connect(m_rightModel, &QAbstractItemModel::rowsRemoved, this, [this] {
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), m_joinedRoles);
+    });
+
+    connect(m_rightModel, &QAbstractItemModel::rowsInserted, this, [this] {
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), m_joinedRoles);
+    });
+
+    connect(this, &QAbstractItemModel::dataChanged, this,
+            [this](auto& topLeft, auto& bottomRight, auto& roles ) {
+        if (roles.contains(m_leftModelJoinRole))
+            emit dataChanged(topLeft, bottomRight, m_joinedRoles);
+    });
+
+    QIdentityProxyModel::setSourceModel(m_leftModel);
+}
+
+QVariant LeftJoinModel::data(const QModelIndex& index, int role) const
+{
+    auto idx = m_leftModel->index(index.row(), 0);
```

Comment:
Probably it would work, but the received index belongs to the join model, not left model. Basically for a model, indexes of that models should be used. Index is quite complex thing, containing also pointer to the model itself. For that reason I would of for not mixing indexes between models.

Reply:
> Index is quite complex thing, containing also pointer to the model itself. 

This is why I was expecting for the `index` to be used (along with `mapToSource`). Using just the row drops the support for columns and parent index.

Since we need this proxy for basic list models and this would definitely work. Moreover, I'm not sure how the proxy will behave when providing table or tree models .

If we're going to support only list models, I'd override the `columnCount` and set it to 0.

If you want to use the received index and add support for columns and tree models I think we need to connect to column change events in the `initialize` (not sure it's enough though)

Reply:
I added passing column here, but especially support for trees is totally out of scope here imo. If needed we can focus on that as well :)